src/zkevm_specs/state.py

```diff
@@ -400,6 +480,20 @@ def get_keys_compressed_in_order(row: Row) -> List[FQ]:
     if row.is_write == 0 and all_keys_eq(row, row_prev):
         assert row.value == row_prev.value
 
+    # 7. Increment mpt_counter
+    #
+    # When row is Storage or Account, increment the mpt_counter by
+    # one, otherwise maintain the same value
+    if row.tag() != Tag.Start:
+        if row.tag() == Tag.Storage or row.tag() == Tag.Account:
+            assert row.mpt_counter == row_prev.mpt_counter + 1
+        else:
+            assert row.mpt_counter == row_prev.mpt_counter
+
+    # 8. RWC !=0 except for Tag.Start
+    if row.tag() != Tag.Start:
+        assert row.rw_counter != 0
```

Comment:
the comment in https://github.com/appliedzkp/zkevm-specs/pull/200/files#diff-8c8275ac52d72de465920627c9e900c9ae5d4fc1aba1e75bd44e5d6c346a4d55L209 suggests that originally we were going to allow rw_counter to be 0 for other tags as well. Do we no longer need that because we're doing MPT lookups for every Storage and Account row?

Will we need to get rid of this constraint when we do the optimization mentioned here: https://github.com/appliedzkp/zkevm-specs/pull/200/files#diff-8c8275ac52d72de465920627c9e900c9ae5d4fc1aba1e75bd44e5d6c346a4d55R279?

Reply:
> the comment in https://github.com/appliedzkp/zkevm-specs/pull/200/files#diff-8c8275ac52d72de465920627c9e900c9ae5d4fc1aba1e75bd44e5d6c346a4d55L209 suggests that originally we were going to allow rw_counter to be 0 for other tags as well. Do we no longer need that because we're doing MPT lookups for every Storage and Account row?

I think originally the only rows (different than Start) where we were thinking of using `rwc = 0` where for Account and Storage to initialize the value (so that we can always get a valid `value_prev`).  A row with `rwc = 0` like that is not tracked by the EVM circuit, and as such an attacker could insert malicious rows anywhere in the State circuit; so we'd need to check that `rwc = 0` can only be enabled in specific scenarios, and otherwise it must be `rwc != 0`.  But then I realized that we already have a `commited_value` column which allows us to always have a valid `value_prev`, so we can remove the need of these Account/Storage rows with `rwc = 0`, and then the rules for `rwc` become simpler.

So removing the Storage/Account rows with `rwc = 0` is unrelated to doing one/many MPT lookup for every Storage/Account.

> Will we need to get rid of this constraint when we do the optimization mentioned here: https://github.com/appliedzkp/zkevm-specs/pull/200/files#diff-8c8275ac52d72de465920627c9e900c9ae5d4fc1aba1e75bd44e5d6c346a4d55R279?

Even with the mentioned optimization we can still keep the same `rwc` rules simplification as in this PR.  The optimization could be implemented by doing the MPT lookup with `(value = value, prev_value = committed_value)` only when the next row keys change (In `committed_value` we have the first read value, and in `value` of the last row for a set of equal keys we have the last write value).

Does this make sense?